cl/phase1/forkchoice/get_head.go

```diff
@@ -43,48 +46,92 @@ func (f *ForkChoiceStore) accountWeights(votes, weights map[libcommon.Hash]uint6
 	return
 }
 
-func (f *ForkChoiceStore) GetHead() (libcommon.Hash, uint64, error) {
+const (
+	sampleFactor = 100
+	sampleBasis  = 80
+)
+
+func (f *ForkChoiceStore) computeVotes(justifiedCheckpoint solid.Checkpoint, checkpointState *checkpointState, auxilliaryState *state.CachingBeaconState) map[libcommon.Hash]uint64 {
+	votes := make(map[libcommon.Hash]uint64)
+	// make an rng generator
+	gen := rand.New(rand.NewSource(time.Now().UnixNano()))
+	if auxilliaryState != nil {
+		startIdx := 0
+		step := 1
+		if f.probabilisticHeadGetter {
+			startIdx = gen.Intn(sampleBasis)
+			step = sampleBasis + gen.Intn(sampleFactor)
+		}
+		for validatorIndex := startIdx; validatorIndex < len(f.latestMessages); validatorIndex += step {
+			message := f.latestMessages[validatorIndex]
+			v := auxilliaryState.ValidatorSet().Get(validatorIndex)
+			if !v.Active(justifiedCheckpoint.Epoch) || v.Slashed() {
+				continue
+			}
+			if _, hasLatestMessage := f.getLatestMessage(uint64(validatorIndex)); !hasLatestMessage || f.isUnequivocating(uint64(validatorIndex)) {
+				continue
+			}
+			votes[message.Root] += v.EffectiveBalance()
+		}
+		boostRoot := f.proposerBoostRoot.Load().(libcommon.Hash)
+		if boostRoot != (libcommon.Hash{}) {
+			boost := auxilliaryState.GetTotalActiveBalance() / auxilliaryState.BeaconConfig().SlotsPerEpoch
+			votes[boostRoot] += (boost * auxilliaryState.BeaconConfig().ProposerScoreBoost) / 100
+		}
+	} else {
```

Comment:
Is the `else` case still useful, or are we just keeping it for tests?

Reply:
yes it is - it fall backs to it also when head state is not avaiable. this forkchoice module so it is not really "consensus"